File: Python/N-queen problem.py
# Function to check if placement is safe
def is_safe(board, curr_row, curr_col):
    for i in range(len(board)):
        placed_row = board[i]
        placed_col = i + 1
        
        # Check diagonals
        if abs(placed_row - curr_row) == \
           abs(placed_col - curr_col):
            return False # Not safe
    return True # Safe to place

# Recursive utility to solve N-Queens
def nqueen_util(col, n, board, result, visited):
  
    # If all queens placed, add to result
    if col > n:
        result.append(board.copy())
        return
      
    # Try each row in column
    for row in range(1, n+1):
      
        # If row not used
        if not visited[row]:
          
            # Check safety
            if is_safe(board, row, col):
              
                # Mark row
                visited[row] = True
                
                # Place queen
                board.append(row)
                
                # Recur for next column
                nqueen_util(col+1, n, board, result, visited)
                
                # Backtrack
                board.pop()
                visited[row] = False

# Main N-Queen solver
def nqueen(n):
    result = []
    board = []
    visited = [False] * (n +1)
    nqueen_util(1, n, board, result, visited)
    return result

# Main function
if __name__ == "__main__":
    n = 4
    result = nqueen(n)
    for res in result:
        print("[", end="")
        for i in range(len(res)):
            print(res[i], end="")
            if i != len(res)-1:
                print(", ", end="")
        print("]")

File: CPP/algorithms/recursion/n_queen_problem.cpp
#include <bits/stdc++.h>
using namespace std;

/*
=====================================================
Algorithm: N-Queens Problem using Backtracking
=====================================================

The N-Queens problem is to place N queens on an N×N chessboard
such that no two queens attack each other. A queen can attack
horizontally, vertically, and diagonally.

Approach:
1. Place queens one column at a time (from left to right).
2. For each column, try placing a queen in every row:
   - Check if the position is safe (no conflict in row, upper-left diagonal, lower-left diagonal).
   - If safe, place the queen and recursively attempt to place the rest.
   - If placing the queen leads to no solution, backtrack (remove the queen and try next row).
3. Stop when all queens are successfully placed.

=====================================================
Time Complexity:
- For each column, we try placing a queen in N rows.
- At each step, we perform an O(N) check for safety (row + 2 diagonals).
- Worst-case time complexity: O(N!) (factorial growth).
- More precisely: O(N * N!) due to the safety check overhead.

Space Complexity:
- O(N^2) for the chessboard representation.
- O(N) recursive call stack (depth = number of columns).
=====================================================
*/

// Function to print the solution board
void printSolution(vector<vector<int>>& board) {
    int n = board.size();
    for (int row = 0; row < n; row++) {
        for (int col = 0; col < n; col++) {
            if (board[row][col] == 1)
                cout << "Q ";  // Queen placed
            else
                cout << ". ";  // Empty cell
        }
        cout << "\n";
    }
}

// Check if it's safe to place a queen at board[row][col]
bool isSafe(vector<vector<int>>& board, int row, int col) {
    int n = board.size();

    // Check row on the left
    for (int j = 0; j < col; j++)
        if (board[row][j] == 1)
            return false;

    // Check upper-left diagonal
    for (int i = row, j = col; i >= 0 && j >= 0; i--, j--)
        if (board[i][j] == 1)
            return false;

    // Check lower-left diagonal
    for (int i = row, j = col; i < n && j >= 0; i++, j--)
        if (board[i][j] == 1)
            return false;

    return true; // Safe to place queen
}

// Recursive utility function to solve N-Queens
bool solveNQUtil(vector<vector<int>>& board, int col) {
    int n = board.size();

    // Base case: all queens are placed
    if (col >= n)
        return true;

    // Try placing queen in each row of current column
    for (int row = 0; row < n; row++) {
        if (isSafe(board, row, col)) {
            // Place queen
            board[row][col] = 1;

            // Recur to place rest of the queens
            if (solveNQUtil(board, col + 1))
                return true;

            // Backtrack: remove queen
            board[row][col] = 0;
        }
    }

    // No valid placement in this column
    return false;
}

// Solve N-Queens problem
bool solveNQ(int n) {
    vector<vector<int>> board(n, vector<int>(n, 0));

    if (!solveNQUtil(board, 0)) {
        cout << "Solution does not exist";
        return false;
    }

    printSolution(board);
    return true;
}

// Driver code
int main() {
    int n = 4;
    solveNQ(n);
    return 0;
}

